### PR TITLE
module 4

### PR DESCRIPTION
**Pull request by NadzeyaZmushka**

### CODE REVIEW

File: repository/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java [hunk 1]
```diff
@@ -118,4 +119,24 @@ public Long count() {
                 .getSingleResult();
     }
 
+    private <T> void prepareSearchQuery(AbstractQuery<T> query, Root<Certificate> root, List<String> tagNames, String name) {
+        CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
+        List<Predicate> predicates = new ArrayList<>();
+        if (!StringUtils.isBlank(name)) {
```
Comment (HannaVatlina) [inline repository/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java]: it is better to use the StringUtils.isNotBlank method

File: repository/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java [hunk 2]
```diff
@@ -118,4 +119,24 @@ public Long count() {
...
+            Predicate predicateForName = criteriaBuilder.like(root.get("name"), "%" + name + "%");
+            predicates.add(predicateForName);
+        }
+        if (tagNames != null && tagNames.size() != 0) {
```
Comment (HannaVatlina) [inline repository/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java]: Please, look into the CollectionUtils#isNotEmpty

File: service/src/test/java/com/epam/esm/service/impl/TagServiceImplTest.java [hunk 3]
```diff
@@ -141,25 +140,25 @@ void testShouldReturnListOfTagsWithNames() {
...
+//    @Test
+//    void testShouldReturnWidelyUsedTag() {
+//        //given
+//        Tag tag = new Tag(1L, "popular");
+//        when(tagRepository.findMostPopularTag(, )).thenReturn(List.of(tag));
+//        //when
+//        List<Tag> actual = tagService.findWidelyUsed(, );
+//        //then
+//        assertEquals(0, actual.size());
+//    }
```
Comment (HannaVatlina) [inline service/src/test/java/com/epam/esm/service/impl/TagServiceImplTest.java]: Please, try not to commit the commented out code. If test doesn't need just remove it, or use @Ignore annotation.

File: web/src/main/java/com/epam/esm/configuration/AuthorizationServerConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                .accessTokenValiditySeconds(180)
+                .refreshTokenValiditySeconds(86400)
+                .scopes("any")
+                .autoApprove(true)
+                .authorities("ADMIN", "USER")
+                .authorizedGrantTypes("password", "refresh_token")
+                .resourceIds("api");
+    }
+
+    //настраиваем свойства и расширенную функциональность конечных точек сервера авторизации
```
Comment (HannaVatlina) [inline web/src/main/java/com/epam/esm/configuration/AuthorizationServerConfiguration.java]: try to use the comment in english

File: web/src/main/java/com/epam/esm/security/UserDetailServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+@Service
+@RequiredArgsConstructor
+public class UserDetailServiceImpl implements UserDetailsService {
+
+    private final UserRepositoryImpl userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        User user = userRepository.findByEmail(username)
+                .orElseThrow(() -> new UsernameNotFoundException("User not found"));
```
Comment (HannaVatlina) [inline web/src/main/java/com/epam/esm/security/UserDetailServiceImpl.java]: "User not found" - need add localization

File: web/src/main/java/com/epam/esm/security/jwt/CustomTokenEnhancer.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    private final UserRepository userRepository;
+
+    @Override
+    public OAuth2AccessToken enhance(OAuth2AccessToken accessToken, OAuth2Authentication authentication) {
+        User user = userRepository.findByEmail(authentication.getName()).orElseThrow(() ->
+                new UsernameNotFoundException(authentication.getName()));
+        DefaultOAuth2AccessToken enhanced = new DefaultOAuth2AccessToken(accessToken);
+        Map<String, Object> extraInfo = new HashMap<>();
+        extraInfo.put("user_id", user.getId());
+        extraInfo.put("iss", "http://localhost:8080");
```
Comment (HannaVatlina) [inline web/src/main/java/com/epam/esm/security/jwt/CustomTokenEnhancer.java]: Please, move "http://localhost:8080" to the configuration

File: repository/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java (hunk 1 above)
Comment (NadzeyaZmushka) [reply repository/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java]: Right, thanks!

File: repository/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java (hunk 2 above)
Comment (NadzeyaZmushka) [reply repository/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java]: Done

File: service/src/test/java/com/epam/esm/service/impl/TagServiceImplTest.java (hunk 3 above)
Comment (NadzeyaZmushka) [reply service/src/test/java/com/epam/esm/service/impl/TagServiceImplTest.java]: Removed

File: web/src/main/java/com/epam/esm/security/UserDetailServiceImpl.java (hunk 5 above)
Comment (NadzeyaZmushka) [reply web/src/main/java/com/epam/esm/security/UserDetailServiceImpl.java]: Done

File: web/src/main/java/com/epam/esm/security/jwt/CustomTokenEnhancer.java (hunk 6 above)
Comment (NadzeyaZmushka) [reply web/src/main/java/com/epam/esm/security/jwt/CustomTokenEnhancer.java]: Done

**Pull request merged**